File: ts/connectors/elk.ts
/**
 * fetch_logs: live query against the hosted logging ELK. Logs are NEVER
 * ingested — queried where they live, with hard caps (context bombs).
 */

import { type DcClient, type Fetcher, makeClient, postJson } from "./auth.js";

export const MAX_HITS = 50;
export const MESSAGE_MAX_CHARS = 400;

/**
 * A bounded, index-scoped `_search` (`size: 0` — proves the query executes
 * and auth is accepted without returning any hit data) against the same
 * configured index and production query shape `fetchLogs()` actually uses.
 * Deliberately NOT `/_cluster/health`: that commonly requires
 * cluster-monitor privileges an index-scoped search credential need not
 * have, so it can fail (or pass) independently of whether the real EIL
 * search path works. Used only by `eil doctor`.
 */
export async function doctorProbe(client: DcClient, env: NodeJS.ProcessEnv): Promise<void> {
  const index = env.EIL_ELK_INDEX ?? "logs-*";
  await postJson(client, `/${index}/_search`, { size: 0, query: { match_all: {} } }, "query");
}

export class ElkClient {
  readonly client: DcClient;
  readonly defaultIndex: string;

  constructor(baseUrl?: string, token?: string, fetcher?: Fetcher) {
    this.client = makeClient("ELK", baseUrl, token, fetcher);
    this.defaultIndex = process.env.EIL_ELK_INDEX ?? "logs-*";
  }

  async fetchLogs(
    query: string,
    index?: string,
    minutes = 60,
    limit = 20,
  ): Promise<Record<string, unknown>> {
    const size = Math.min(limit, MAX_HITS);
    const data = await postJson(
      this.client,
      `/${index ?? this.defaultIndex}/_search`,
      {
        size,
        sort: [{ "@timestamp": "desc" }],
        query: {
          bool: {
            must: [{ query_string: { query } }],
            filter: [{ range: { "@timestamp": { gte: `now-${minutes}m` } } }],
          },
        },
      },
      "query", // Elasticsearch _search is a read — safe to retry
    );
    const hits = data.hits ?? {};
    const results = (hits.hits ?? []).map((h: any) => {
      const src = h._source ?? {};
      return {
        ts: src["@timestamp"],
        level: src.level ?? src["log.level"],
        service: src.service ?? src.kubernetes?.container?.name,
        message: String(src.message ?? "").slice(0, MESSAGE_MAX_CHARS),
      };
    });
    const total = hits.total;
    return {
      query,
      window_minutes: minutes,
      total: typeof total === "object" && total !== null ? total.value : (total ?? results.length),
      results,
    };
  }
}

File: src/eil/connectors/elk.py
"""fetch_logs: live query against the hosted logging ELK.

Logs are NEVER ingested into the knowledge plane (volume, retention, cost) —
this queries them where they live, with hard caps because logs are the
classic context bomb. Env: EIL_ELK_URL, EIL_ELK_TOKEN (or EIL_ELK_USER for
basic auth), EIL_ELK_INDEX default pattern.
"""

from __future__ import annotations

import os
from typing import Any

from eil.connectors.auth import make_client

MAX_HITS = 50
MESSAGE_MAX_CHARS = 400


class ElkClient:
    def __init__(self, base_url: str | None = None, token: str | None = None) -> None:
        self.base_url = (base_url or os.environ["EIL_ELK_URL"]).rstrip("/")
        self.http = make_client("ELK", self.base_url, token)
        self.default_index = os.environ.get("EIL_ELK_INDEX", "logs-*")

    def fetch_logs(
        self, query: str, index: str | None = None, minutes: int = 60, limit: int = 20
    ) -> dict[str, Any]:
        limit = min(limit, MAX_HITS)
        resp = self.http.post(
            f"/{index or self.default_index}/_search",
            json={
                "size": limit,
                "sort": [{"@timestamp": "desc"}],
                "query": {
                    "bool": {
                        "must": [{"query_string": {"query": query}}],
                        "filter": [{"range": {"@timestamp": {"gte": f"now-{minutes}m"}}}],
                    }
                },
            },
        )
        resp.raise_for_status()
        hits = resp.json().get("hits", {})
        results = []
        for h in hits.get("hits", []):
            src = h.get("_source", {})
            results.append(
                {
                    "ts": src.get("@timestamp"),
                    "level": src.get("level") or src.get("log.level"),
                    "service": src.get("service") or src.get("kubernetes", {}).get("container", {}).get("name"),
                    "message": str(src.get("message", ""))[:MESSAGE_MAX_CHARS],
                }
            )
        total = hits.get("total", {})
        return {
            "query": query,
            "window_minutes": minutes,
            "total": total.get("value", len(results)) if isinstance(total, dict) else total,
            "results": results,
        }
